packages/flutter/lib/src/widgets/restoration.dart

```diff
@@ -0,0 +1,1031 @@
+/// [initWithValue] repeats. Whenever [initWithValue] is called, the property
+/// should forget the old value it was wrapping and re-initialize itself with
+/// the newly provided value.
+///
+/// In a typical use case, a subclass of [RestorableProperty] is instantiated
+/// either to initialize a member variable of a [State] object or within
+/// [State.initState]. It is then registered to a [RestorationMixin] in
+/// [RestorationMixin.restoreState] and later [dispose]ed in [State.dispose].
+/// For less common use cases (e.g. if the value stored in a
+/// [RestorableProperty] is only needed while the [State] object is in a certain
+/// state), a [RestorableProperty] may be registered with a [RestorationMixin]
+/// any time after [RestorationMixin.restoreState] has been called for the first
+/// time. A [RestorableProperty] may also be unregistered from a
+/// [RestorationMixin] before the owning [State] object is disposed by calling
+/// [RestorationMixin.unregisterFromRestoration]. This is uncommon, though, and
+/// will delete the information that the property contributed from the
+/// restoration data (meaning the value of the property will no longer be
+/// restored during a future state restoration).
+///
+/// See also:
+///
+///  * [RestorableValue], which is a [RestorableProperty] that makes the wrapped
+///    value accessible to the owning [State] object via a `value`
+///    getter and setter.
+///  * [RestorationMixin], to which a [RestorableProperty] must be registered.
+///  * [RestorationManager], which describes how state restoration works in
+///    Flutter.
+abstract class RestorableProperty<T> extends ChangeNotifier {
+  /// Called by the [RestorationMixin] with the `value` returned by either
+  /// [createDefaultValue] or [fromPrimitives] to set the value that this
+  /// property currently wraps.
+  ///
+  /// The [initWithValue] method may be called multiple times throughout the
+  /// life of the [RestorableProperty] whenever new restoration data has been
+  /// provided to the [RestorationMixin] the property is registered to. When
+  /// [initWithValue] is called, the property should forget its previous value
+  /// and re-initialize itself to the newly provided `value`.
+  void initWithValue(T value);
+
+  /// Called by the [RestorationMixin] if no restoration data is available to
+  /// restore the value of the property from to obtain the default value for the
+  /// property.
+  ///
+  /// The method returns the default value that the property should wrap if no
+  /// restoration data is available. After this method has been called,
+  /// [initWithValue] will be invoked with the returned value.
+  ///
+  /// The method may be called multiple times throughout the life of the
+  /// [RestorableProperty]. Whenever new restoration data has been provided to
+  /// the [RestorationMixin] the property is registered to, either this method
+  /// or [fromPrimitives] is called before [initWithValue] is invoked.
```

Comment:
might be clearer to say something like "after this is called, initWithValue will be called with this method's return value" or "the value return by this method is then passed to initWithValue", and then separately say "see also fromPrimitives which is called when ... instead"